Merge MutableGlobalCache into GlobalCache

knelli2 opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
clang-format MockRuntimeSystem

And ParallelComponentHelpers

knelli2 committed:
Merge MutableGlobalCache into GlobalCache

The mutable part of the cache is now once per charm-node, not once
per charm-core as it was before

wthrowe commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -634,16 +440,48 @@ template <typename GlobalCacheTag, typename Function, typename... Args>
+  // list (if a callback calls mutable_cache_item_is_ready) will be
+  // saved and will not be invoked here.
+  for (auto& callback : callbacks) {
+    callback->invoke();
```
I think you need to hold the write lock for this entire function.  Here's the potential problem I see:
* We have two writers, A and B preparing to make independent changes to a cache object.
* We have an element E with a callback waiting for the change B is going to make.

Suppose the following sequence of events:
1. A mutates the object, copies the callback list, and starts the callback for element E.
2. E checks the current value and determines it is not ready.
3. B mutates the object, copies the empty callback list, and returns with nothing to do.
4. E returns a new callback, which is added to the callback list.
5. A returns.

We now have E waiting on a change that has already happened.

wthrowe commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -634,16 +440,48 @@ template <typename GlobalCacheTag, typename Function, typename... Args>
+          const std::lock_guard<std::mutex> lock(mutex);
+          Function::apply(make_not_null(&(std::get<0>(
+                              tuples::get<tag>(mutable_global_cache_)))),
+                          local_args...);
```
Should we dereference `unique_ptr`s here?  I guess not doing so is consistent with `db::mutate`, but you can't actually change what is being pointed to since that would violate the always-readable guarantee.

nilsdeppe replied:
I'd be in favor of dereferences so that people can't change things they shouldn't be able to

wthrowe commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -865,17 +667,21 @@ auto get_parallel_component(const GlobalCache<Metavariables>& cache)
+    // because we require all mutable tags to be able to be read at all times
+    // (even when being written to)
+    if constexpr (tt::is_a_v<std::unique_ptr, typename tag::tag::type>) {
+      return *std::get<0>((tuples::get<tag>(cache.mutable_global_cache_)));
```
Nit: extra parens compared to two lines down.

wthrowe commented on tests/Unit/Parallel/Test_GlobalCache.cpp:
```diff
@@ -562,28 +551,14 @@ void Test_GlobalCache<Metavariables>::exit_if_done(int index) {
-PUP::able::PUP_ID Arthropod::my_PUP_ID = 0;  // NOLINT
+    // clang-tidy: possibly throwing constructor static storage
+    // clang-tidy: false positive: redundant declaration
+    PUP::able::PUP_ID Triangle::my_PUP_ID = 0;  // NOLINT
```
clang-format is dumb.  Don't let it make the formatting worse.

wthrowe commented on tests/Unit/Parallel/Test_GlobalCache.cpp:
```diff
@@ -228,13 +228,24 @@ class UseCkCallbackAsCallback : public Parallel::Callback {
   CkCallback callback_;
 };
 
+// We have to define our expected mutable values statically out here because the
+// mutable part of the cache is once per node and we are only running on one
+// node. This means that when we run `run_test_one` on the first array element
+// and it mutates a mutable tag, that tag gets immediately changed inside the
+// mutable part of the cache. Then when the second array element runs
+// `run_test_one` this value has already been mutated, so we need to keep track
+// of these changes outside the functions with these static vars.
+static double expected_weight = 160.0;
+static std::string expected_email = "joe@somewhere.com";
+static size_t expected_number_of_legs = 6;
```
Can you explain in a github comment what is different from the old behavior that requires this?

knelli2 replied:
So I originally left this part of the test alone, but found that it failed. Here is the flow of the `run_test_one` function:

1. Check value of tag is expected *starting* value in `run_test_one`
2. Mutate tag in `run_test_one`
3. Immediately call `run_test_two` inlined.

Now this is run on every element of the array parallel component (4 elements). What this means is that the mutable tag will be edited 4 times because `run_test_one` is run on all array elements (not a big deal because it's being mutated to the same value). It appears that in the old test, possibly because the mutate went through a proxy to the const cache and then a proxy to the mutable cache, the actual mutate call wasn't run until all 4 `run_test_one` functions were called. This means that step 1., the check that the tag is its *starting* value, always passed. However, now with my changes the mutate only goes through one proxy call to the cache. Apparently this caused the actual mutate to be run before all 4 of the `run_test_one` functions could run, so step 1. failed in at least one `run_test_one` because the tag is no longer at its starting value.

That's why I added the static variables. To keep track of the actual current state of the tags so that they can be compared in all `run_test_X` whenever they are run. I think the test before just got lucky that that's how the execution order worked out.

wthrowe replied:
I think that breaks a lot of the purpose of the test.  All the `mutable_cache_item_is_ready` calls are trivially being satisfied if everything is inline, right?  I don't think this is testing registering and calling callbacks anymore.

It might be necessary to have the different threads do different things to make them they actually have to wait on one another.

knelli2 replied:
What if I leave the functions as they originally were except I remove the calls to the other functions inside them. Then I call each one on a different element? Like this:

```cpp
array_proxy[3].run_test_five();
array_proxy[2].run_test_four();
array_proxy[1].run_test_three();
array_proxy[0].run_test_two();
array_proxy[3].run_test_one();
```

Since `run_test_one` is what does the mutate that kicks everything off, a bunch of callbacks should be registered and called.

wthrowe replied:
Yes, that sounds good.  I don't see any reason they have to be chained.

wthrowe commented on tests/Unit/Parallel/Test_GlobalCache.cpp:
```diff
@@ -290,18 +303,23 @@ void TestArrayChare<Metavariables>::run_test_two() {
           *Parallel::local_branch(global_cache_proxy_),
```
Comment a few lines up still refers to "150".  Haven't looked for other similar things.

kidder commented on src/Parallel/GlobalCacheDeclare.hpp:
```diff
@@ -2,9 +2,7 @@
-/// Forward-declares CProxy_GlobalCache which MutableGlobalCache needs, but
-/// GlobalCache is defined after MutableGlobalCache. Also forward declares
-/// ResourceInfo which the GlobalCache has an entry method for.
+/// Forward-declares ResourceInfo which the GlobalCache has an entry method for.
```
is the forward declaration of CProxy_GlobalCache still needed?

knelli2 replied:
Yes I got errors if I didn't have it and didn't bother trying to track things down.

kidder commented on src/Parallel/Main.hpp:
```diff
@@ -479,10 +470,9 @@ Main<Metavariables>::Main(CkArgMsg* msg) {
   // GlobalCache before we pass it to any other parallel component rather than
-  // having the tags in the MutableGlobalCache and using a mutate call to set
-  // them.
+  // having the a mutable tag and using a mutate call to set them.
```
the a -> a

kidder commented on src/ParallelAlgorithms/Interpolation/InterpolationTarget.hpp:
```diff
@@ -145,9 +145,9 @@ namespace intrp {
-///   `domain::Tags::FunctionsOfTime` in MutableGlobalCache, so that
-///   once `FunctionsOfTime` are updated, VerifyTemporalIdsAndSendPoints
-///   is called.
+///   `domain::Tags::FunctionsOfTime` in mutable part of the GlobalCache, so
```
in the mutable

kidder commented on src/ParallelAlgorithms/Interpolation/InterpolationTarget.hpp:
```diff
@@ -251,10 +251,8 @@ namespace intrp {
-/// copy per core, rather than one copy per Element), or even in the
-/// GlobalCache (one copy per node) since the points need be computed only
-/// once.
+/// be to place the points in  the GlobalCache (one copy per node) since the
```
remove extra space between in and the

nilsdeppe commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -620,12 +402,36 @@ template <typename Metavariables>
+    // Second mutex is for vector of callbacks
+    std::mutex& mutex = tuples::get<MutexTag<tag>>(mutexes_).second;
+    {
+      // Scoped for lock guard
```
Since this is all inside an `if` why not do:
```cpp
const std::lock_guard<std::mutex> lock(tuples::get<MutexTag<tag>>(mutexes_).second);
...
```
and avoid the extra indentation?

knelli2 replied:
Does the `return` affect this? Meaning that other destructors have to be called in addition to the lock guard destructor? Which order are they called and could that affect performance?

nilsdeppe replied:
destructors are called in the opposite order of constructors, and are called before returning from the function. In terms of performance, there won't be a difference between:
```cpp
}
return blah;
```
and
```cpp
return blah;
```
The machine code should be identical in those two cases.

Does that answer your questions?

knelli2 replied:
Yup! Thank you!

nilsdeppe reviewed (commented):
LGTM, couple minor suggestions.

knelli2 commented:
@nilsdeppe @wthrowe @kidder Rebased and added a fixup commit. Also changed this from a draft to a real pull request because I believe we have the thread-safety things figured out.

nilsdeppe reviewed (commented):
I'm happy. No need to have me look again before merging!

wthrowe commented on tests/Unit/Parallel/Test_GlobalCache.cpp:
```diff
@@ -220,32 +220,25 @@ struct TestMetavariables {
+    p | callback_;
+    // ERROR(
+    //     "Should not be pupping a UseCkCallbackAsCallback, since it is always
+    //     " "stored in the local GlobalCache");
```
Remove?

wthrowe reviewed (changes requested):
Otherwise looks good.

knelli2 commented:
Had to fix just a couple more things since #5065 got merged (just `Test_Evolution.cpp` and `Test_DgDomain.cpp`)
